cubic-server/protocol_id_converter/blockStates.cpp

```diff
@@ -0,0 +1,88 @@
+
+void Blocks::GlobalPalette::initialize(std::string path) {
+    if (!std::filesystem::exists(path)) {
+        LERROR("File " << path << " not found !");
+        return;
+    }
+    nlohmann::json file = nlohmann::json::parse(std::ifstream(path));
+    for(auto block : file.items()) {
+        Blocks::InternalBlock b;
+        b.name = block.key();
+        if (block.value().size() == 2) {
+            int maxThingy = 0; // Sorry for the name I don't how to call it
+            int wheight = 1;
+            for (auto property : block.value()["properties"].items()) {
+                Blocks::InternalProperty p;
+                p.name = property.key();
+                p.baseWeight = wheight;
+                for (auto value : property.value()) {
+                    p.values.push_back(value);
+                }
+                maxThingy += wheight * (p.values.size() - 1);
+                wheight = maxThingy + 1;
+                b.properties.push_back(p);
+            }
+        }
+        b.baseProtocolId = block.value()["states"][0]["id"];
+        b.maxProtocolId = block.value()["states"][block.value()["states"].size() - 1]["id"];
+        this->_blocks.push_back(b);
+    }
+}
+
+Blocks::BlockId Blocks::GlobalPalette::fromBlockToProtocolId(Blocks::Block block) const {
+    auto internalBlock = std::find_if(this->_blocks.begin(), this->_blocks.end(), [&block](Blocks::InternalBlock b) {
+        return b.name == block.name;
+    });
+    if (internalBlock == this->_blocks.end()) {
+        LERROR("Block not found in palette (name: " << block.name << ")");
+        return 0;
+    }
+    if (block.properties.size() != internalBlock->properties.size()) {
+        LERROR("Block properties size mismatch with block " << block.name);
+        return 0;
+    }
+    if (block.properties.size() == 0 && internalBlock->properties.size() == 0)
+        return internalBlock->baseProtocolId;
+
+    BlockId id = internalBlock->baseProtocolId;
+    for (auto property : block.properties) {
+        auto internalProperty = std::find_if(internalBlock->properties.begin(), internalBlock->properties.end(), [&property](Blocks::InternalProperty p) {
+            return p.name == property.first;
+        });
+        if (internalProperty == internalBlock->properties.end()) {
+            LERROR("Property not found (name: " << property.first << ")");
+            return 0;
+        }
+        auto value = std::find(internalProperty->values.begin(), internalProperty->values.end(), property.second);
+        if (value == internalProperty->values.end()) {
+            LERROR("Value not found (name: " << property.second << ")");
+            return 0;
+        }
+        id += internalProperty->baseWeight * (value - internalProperty->values.begin());
+    }
+    return id;
+}
+
+Blocks::Block Blocks::GlobalPalette::fromProtocolIdToBlock(Blocks::BlockId id) const {
+    for (auto b : this->_blocks) {
+        if (id >= b.baseProtocolId && id <= b.maxProtocolId) {
```

Comment:
```suggestion
        if (id < b.baseProtocolId || id > b.maxProtocolId)
                continue;
```

Reply:
Isn't this dangerous to let a return keyword in the for loop without having the if scope ?

Reply:
Nop